File: Data Structures & Algorithms/longest-consecutive-sequence/submission-1.cpp
class Solution {
public:
    int longestConsecutive(vector<int>& nums) {
        unordered_map<int, int> m;
        int res = 0;
        for(int i = 0; i < nums.size(); i++)
        {
            if(m[nums[i]]) continue;
            m[nums[i]] = 1;
            m[nums[i]] += m[nums[i]-1] + m[nums[i]+1];
            if(m[nums[i]-1])
            {
                m[nums[i]-m[nums[i]-1]] = m[nums[i]];
            }
            if(m[nums[i]+1])
            {
                m[nums[i]+m[nums[i]+1]] = m[nums[i]];
            }
            res = max(res, m[nums[i]]);
        }
        return res;
    }
};

File: Data Structures & Algorithms/trapping-rain-water/submission-1.py
class Solution:
    def trap(self, height: List[int]) -> int:
        # at each point we want to find the highest to the left and highest to the right (inclusive)
        # then we subtract the value of every height we come across
        mh = []
        cmax = 0
        for h in height:
            cmax = max(h, cmax)
            mh.append(cmax)
        cmax = 0
        res = 0
        for i in range(len(height) - 1, -1, -1):
            cmax = max(height[i], cmax)
            res += min(mh[i], cmax)
            res -= height[i]
        return res